extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java

```diff
@@ -0,0 +1,140 @@
+package io.quarkus.infinispan.client.runtime.cache;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import javax.annotation.Priority;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.commons.CacheException;
+import org.jboss.logging.Logger;
+
+import io.quarkus.infinispan.client.CacheResult;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.TimeoutException;
+import io.smallrye.mutiny.Uni;
+
+@CacheResult(cacheName = "") // The `cacheName` attribute is @Nonbinding.
+@Interceptor
+@Priority(CacheInterceptor.BASE_PRIORITY + 2)
+public class CacheResultInterceptor extends CacheInterceptor {
+
+    private static final Logger LOGGER = Logger.getLogger(CacheResultInterceptor.class);
+    private static final String INTERCEPTOR_BINDING_ERROR_MSG = "The Quarkus Infinispan Client extension is not working properly (CacheResult interceptor binding retrieval failed), please create a GitHub issue in the Quarkus repository to help the maintainers fix this bug";
+    private final Map<String, Map<Object, CompletableFuture<Object>>> synchronousGetLocks = new ConcurrentHashMap<>();
```

Comment:
I'd like to confirm something based on this:
```
@ApplicationScoped
class A {
    @CacheResult(cacheName = "foo")
    String getValue(Object key) { /** body omitted */ }
}

@Singleton
class B {
    @CacheResult(cacheName = "foo")
    String getValue(Object key) { /** body omitted */ }
}
```

Unless I misunderstood what @mkouba wrote, if `A#getValue` and `B#getValue` are called concurrently with the same arg, the remote Infinispan cache will _not_ be protected against concurrent read/write access because each bean (A or B) will have its own instance of `CacheResultInterceptor` and therefore its own locks Map. Doesn't that defeat the very reason this Map exists? Or maybe I missed something.

Reply:
Unless an easy solution exists, probably we should consider writing down that limitation/risk and write down some  of good practices. WDYT?

Reply:
The Map could probably be moved to a separate `@ApplicationScoped` or `@Singleton` bean and that bean could then be injected in the interceptors.

Reply:
obviously @gwenneg 🤦‍♀️ ;) thanks will do that

Reply:
I added a change to scope the map into a bean @gwenneg 
let me know if I missed something please